### PR TITLE
WorldMapPlugin: Show quest state

### PR DESCRIPTION
**Pull request by connorjclark**
Found it simpler to restart than rebase #8199.

I'd like to follow up this PR with adding the locations to the Quest enum, and removing the QuestLocation enum.

I think the icon for "In Progress" quests should be yellow, to match the color in the quest log. I just copied the images from the original PR, but I could remake the icon as yellow tomorrow.

There are some quests that are started in the same location. The logic will only show the latest non-completed quest, else it will show the last quest.

image

image

image

Also, I figured rolling this into the existing quest start config option made sense, instead of making a new config option.

### CODE REVIEW

**Comment by Nightfirecat** [comment]
FYI this should be co-authored-by the original PR writer, and should include their copyright on any files they added it to originally since you've used their patch as a starting point.

**Comment by Hoten** [comment]
Good call - I add `Co-authored-by:` to the commit message.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java [hunk 1]
```diff
@@ -27,151 +27,162 @@
...
+	MISTHALIN_MYSTERY(Quest.MISTHALIN_MYSTERY, new WorldPoint(3234, 3155, 0)),
+	PIRATES_TREASURE(Quest.PIRATES_TREASURE, new WorldPoint(3051, 3252, 0)),
+	PRINCE_ALI_RESCUE(Quest.PRINCE_ALI_RESCUE, new WorldPoint(3301, 3163, 0)),
+	THE_RESTLESS_GHOST(Quest.THE_RESTLESS_GHOST, new WorldPoint(3240, 3210, 0)),
+	RUNE_MYSTERIES(Quest.RUNE_MYSTERIES, new WorldPoint(3210, 3220, 0)),
+	SHEEP_SHEARER(Quest.SHEEP_SHEARER, new WorldPoint(3190, 3272, 0)),
+	SHIELD_OF_ARRAV(Quest.SHIELD_OF_ARRAV, new WorldPoint(3208, 3495, 0)),
+	VAMPIRE_SLAYER(Quest.VAMPIRE_SLAYER, new WorldPoint(3096, 3266, 0)),
+	WITCHS_POTION(Quest.WITCHS_POTION, new WorldPoint(2967, 3203, 0)),
+	// X_MARKS_THE_SPOT("X Marks the Spot", new WorldPoint(3227, 3242, 0)),
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: What is this? Missing from `Quests`?

**Comment by Trevor159** [comment]
`X_MARKS_THE_SPOT(550, "X Marks the Spot"),`
should be added to the quest enum

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 2]
```diff
@@ -271,4 +287,57 @@ private void updateShownIcons()
 			}).map(TeleportPoint::new)
 			.forEach(worldMapPointManager::add);
 	}
+
+	private void updateQuestStartPointIcons()
+	{
+		if (!config.questStartTooltips()) return;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: This causes this to not remove the quests from the map when quest start tooltips are toggled off after already showing. That is why it was removing the points before checking the new config. I also don't see a need to move it out of the `updateShowIcons` call.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 3]
```diff
@@ -271,4 +287,57 @@ private void updateShownIcons()
...
+		{
+			quest = quests[quests.length - 1];
+		}
+
+		BufferedImage icon = BLANK_ICON;
+		String tooltip = "";
+		if (quest != null)
+		{
+			tooltip = quest.getName();
+			switch (quest.getState(client))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: This call is required to be on the client thread, I think it will assert fail if you run your client with `-ea`. The config change events can come in from other threads.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 4]
```diff
@@ -271,4 +287,57 @@ private void updateShownIcons()
...
 			.forEach(worldMapPointManager::add);
 	}
+
+	private void updateQuestStartPointIcons()
+	{
+		if (!config.questStartTooltips()) return;
+
+		worldMapPointManager.removeIf(QuestStartPoint.class::isInstance);
+		Arrays.stream(QuestStartLocation.values())
+			.map(value -> createQuestStartPoint(value))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]:
```suggestion
			.map(this::createQuestStartPoint)
```

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java [hunk 5]
```diff
@@ -114,6 +128,15 @@ protected void startUp() throws Exception
 		updateShownIcons();
 	}
 
+	@Subscribe
+	private void onWidgetLoaded(WidgetLoaded l)
+	{
+		if (l.getGroupId() == WidgetID.WORLD_MAP_GROUP_ID)
+		{
+			updateQuestStartPointIcons();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Why is this required?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java (hunk 1 above)
Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: Ah, yes I believe it is. I'll add the enum you provided. Where do you reference the quest id from tho?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 2 above)
Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Got it- can move the point removing back to `updateShowIcons`. However, the rest didn't work when it was in `updateShowIcons`, I was getting an error when accessing the `client` for quest state. So I had to move it to `onWidgetLoaded`

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 5 above)
Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: I got this from the last PR, no idea. Maybe @Rens-br can explain?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 3 above)
Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Ah, interesting. I'll see if it fails as expected later this evening. Hopefully I can find a solution by studying how other plugins access the client.

Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: I tried changing the config during play, but no crash. I did notice that if you have the world map open, and you toggle the WorldMapPlugin, the quest icons won't come back until you re-open the world map. That's because the icons are not done in `startUp`.

Attempting to access client in `startUp` does crash though:

<details>
<summary>expand</summary>

```
2019-06-11 22:00:36 [main] WARN  n.r.client.plugins.PluginManager - Unable to start plugin WorldMapPlugin. {}
net.runelite.client.plugins.PluginInstantiationException: java.lang.reflect.InvocationTargetException
	at net.runelite.client.plugins.PluginManager.startPlugin(PluginManager.java:342)
	at net.runelite.client.plugins.PluginManager.startCorePlugins(PluginManager.java:207)
	at net.runelite.client.RuneLite.start(RuneLite.java:296)
	at net.runelite.client.RuneLite.main(RuneLite.java:221)
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.desktop/java.awt.EventQueue.invokeAndWait(EventQueue.java:1367)
	at java.desktop/java.awt.EventQueue.invokeAndWait(EventQueue.java:1342)
	at java.desktop/javax.swing.SwingUtilities.invokeAndWait(SwingUtilities.java:1480)
	at net.runelite.client.plugins.PluginManager.startPlugin(PluginManager.java:314)
	... 3 common frames omitted
Caused by: java.lang.AssertionError: null
	at client.runScript(client.java:41459)
	at net.runelite.api.Quest.getState(Quest.java:200)
	at net.runelite.client.plugins.worldmap.WorldMapPlugin.createQuestStartPoint(WorldMapPlugin.java:310)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at net.runelite.client.plugins.worldmap.WorldMapPlugin.updateQuestStartPointIcons(WorldMapPlugin.java:299)
	at net.runelite.client.plugins.worldmap.WorldMapPlugin.startUp(WorldMapPlugin.java:129)
	at net.runelite.client.plugins.PluginManager.lambda$startPlugin$3(PluginManager.java:318)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:303)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

</details>

Comment (abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: you need to invoke `Quest::getState` on the client thread via `ClientThread::invoke` because it calls `Client::getVar`, which is not threadsafe.

Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Alrighty, I figured out the threading constraints. Was able to remove the on widget loaded too.

**Comment by Hoten** [comment]
I won't bother with making the "In progress" icon yellow.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 3 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: This is still on the wrong thread. If you run the client with JVM option `-ea` and toggle the quest start tooltips config setting you should be able to see it.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 5 above)
Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: This won't work probably generally since startUp tends to fire before anyone is ever logged in. Hooking to the map widget load like before is probably better since it also handles whenever you change accounts

Comment (Rens-br) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: I can't remember fully but I think it was so it only updates when the worldmap is loaded instead of any widget

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java (hunk 3 above)
Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/WorldMapPlugin.java]: Ah, I see it now. I was only toggling the entire plugin. Done.

**Comment by Hoten** [comment]
Updated images, made it pixel perfect on the minimap, fixed threading issues. should be g2g

File: runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java [hunk 6]
```diff
@@ -27,151 +27,162 @@
 
 import lombok.Getter;
 import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.Quest;
 
+// TODO: add locations to net.runelite.api.Quest, and collect
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: I don't see any reason for `Quest` to contain the start point, would just remove this todo

Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: It collapses these two enums (which are just joined via the quest id) into one.  also makes it so you can't add a new Quest and forget to add a location.

Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: `Quest` is just supposed to allow plugins to uniquely identify quests. The world map plugin cares about the start point so it can add an enum for each quest to contain the extra data it needs - but I don't see why that should be in the core api.

Comment (Hoten) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldmap/QuestStartLocation.java]: I'm actually surprised that other plugins haven't found a reason to use quest locations. Something related to route planning or something. Since nothing else uses it, I appreciate the argument that it shouldn't be in core.

**Comment by Adam-** [comment]
This doesn't work if I log out and log in to a different account. And I don't actually see why it works to begin with. Previously you had it reload the icons on map widget load which I think might work.

**Comment by Adam-** [comment]
I re-added the widget load event and that seemed to work.
